_posts/2020-05-08-coupling-and-cohesion.markdown

```diff
@@ -0,0 +1,138 @@
+---
+layout: post
+title: "Better Software Design With Coupling and Cohesion"
+date: 2020-05-08 17:30:00
+categories: ["learning", "software-development"]
+author: rdormer
+---
+
+One of the most fundamental tasks when writing or refactoring software
+of any kind is breaking the problem down into smaller parts. When
+you're first starting out - and even as you continue to gain
+experience - figuring out what those parts should be, and where they
+should live within a codebase can be a daunting task.  Design patterns
+and principles can help, but trying to keep them in mind as you design
+and implement solutions can be overwhelming. Thankfully, there's a
+pair of principles that can cut many of these gordian knots, and
+render decision making much clearer, simpler, and easier to articulate
+to others. Understanding and using the concepts of *coupling* and
+*cohesion* to guide your design and refactoring decisions yielded
+immediate results for me.
+
+<!--more-->
+
+Judging from a lot of the code that I've had
+to work on over the years, it could help quite a few other developers
+as well.
+
+## The problem in a nutshell
+
+A lot of my early design efforts tended to wind up with amorphously
+defined "layers," catch-all classes full of random junk, and clunky
+abstractions. I knew about a laundry list of design principles like
+SOLID and DRY, [abstraction](https://simpleprogrammer.com/respecting-abstraction)
+and design patterns. I knew that classes should "talk" to each other
+as little as possible. But, when it came time to actually write (or
+improve) code, the results were always somehow less than satisfactory.
+Ideally, we want neatly factored designs with relatively small methods
+and classes, each of which only does one thing, with zero side
+effects. But as I tried to reach that ideal, I seemed to always wind
+up with pieces - sometimes large pieces - of code that didn't
+obviously belong in any particular place. Worse, they couldn't be put
+anywhere without bloating already existing classes or otherwise
+ruining the organization of the code. The question of "where should
+this code live" is something that every working developer will run
+into over and over again.
+
+Learning about the twin concepts of coupling and cohesion suddenly
```

Comment:
Maybe you can include a link to this page? https://wiki.c2.com/?CouplingAndCohesion